# Problem Set 1, due April 13th at 3:30pm 
(in other words, it is due at noon, or one minute after 11:59am).

###Before You Start
Make sure to at least take a basic tutorial in the IPython notebook, otherwise you'll be totally lost.  For this problem set, you should download INFX574-PS1.ipynb and the flights.zip dataset from Canvas. Create a local copy of the notebook and rename it LASTNAME_FIRSTNAME-PS1.ipynb. Then edit your renamed file directly in your browser by typing:
```
ipython notebook <name_of_downloaded_file>
```

You should also make sure the following libraries load correctly (click on the box below and hit Ctrl-Enter)

In [1]:
#IPython is what you are using now to run the notebook
import IPython
print "IPython version:      %6.6s (need at least 1.0)" % IPython.__version__

# Numpy is a library for working with Arrays
import numpy as np
print "Numpy version:        %6.6s (need at least 1.7.1)" % np.__version__

# SciPy implements many different numerical algorithms
import scipy as sp
print "SciPy version:        %6.6s (need at least 0.12.0)" % sp.__version__

# Pandas makes working with data tables easier
import pandas as pd
print "Pandas version:       %6.6s (need at least 0.11.0)" % pd.__version__

# Module for plotting
import matplotlib
print "Mapltolib version:    %6.6s (need at least 1.2.1)" % matplotlib.__version__

# SciKit Learn implements several Machine Learning algorithms
import sklearn
print "Scikit-Learn version: %6.6s (need at least 0.13.1)" % sklearn.__version__

IPython version:       4.1.2 (need at least 1.0)
Numpy version:        1.10.4 (need at least 1.7.1)
SciPy version:        0.17.0 (need at least 0.12.0)
Pandas version:       0.18.0 (need at least 0.11.0)
Mapltolib version:     1.5.1 (need at least 1.2.1)
Scikit-Learn version: 0.17.1 (need at least 0.13.1)


##About the Problem Set: 
This is the same problem set used by Emma Spiro in INFX573. The only difference is that instead of doing the problem set in R, you will use Python and the IPython notebook.

##Instructions: 
In this problem set you will perform a basic exploratory analysis on an example dataset, bringing to bear all of your new skills in data manipulation and visualization. You will be required to submit well commented python code, documenting all code used in this problem set, along with a write up answering all questions below. Use figures as appropriate to support your answers, and when required by the problem. 
This data set uses the NYCFlights13 dataset. You can download the dataset from canvas.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
flights_df= pd.read_csv('/Users/abhi/Documents/UW/Courses/Spring_Quarter_17/INFX-574/data/Problem_Set_1/flights.csv')

In [3]:
print flights_df.shape
print flights_df.columns
print flights_df.dtypes

(336776, 17)
Index([u'Unnamed: 0', u'year', u'month', u'day', u'dep_time', u'dep_delay',
       u'arr_time', u'arr_delay', u'carrier', u'tailnum', u'flight', u'origin',
       u'dest', u'air_time', u'distance', u'hour', u'minute'],
      dtype='object')
Unnamed: 0      int64
year            int64
month           int64
day             int64
dep_time      float64
dep_delay     float64
arr_time      float64
arr_delay     float64
carrier        object
tailnum        object
flight          int64
origin         object
dest           object
air_time      float64
distance        int64
hour          float64
minute        float64
dtype: object


In [4]:
flights_df.dest.unique()
flights_df.head(10)

Unnamed: 0  year  month  day  dep_time  dep_delay  arr_time  arr_delay  \
0           1  2013      1    1     517.0        2.0     830.0       11.0   
1           2  2013      1    1     533.0        4.0     850.0       20.0   
2           3  2013      1    1     542.0        2.0     923.0       33.0   
3           4  2013      1    1     544.0       -1.0    1004.0      -18.0   
4           5  2013      1    1     554.0       -6.0     812.0      -25.0   
5           6  2013      1    1     554.0       -4.0     740.0       12.0   
6           7  2013      1    1     555.0       -5.0     913.0       19.0   
7           8  2013      1    1     557.0       -3.0     709.0      -14.0   
8           9  2013      1    1     557.0       -3.0     838.0       -8.0   
9          10  2013      1    1     558.0       -2.0     753.0        8.0   

  carrier tailnum  flight origin dest  air_time  distance  hour  minute  
0      UA  N14228    1545    EWR  IAH     227.0      1400   5.0    17.0  
1      UA  N24211    1714    LGA  IAH     227.0      1416   5.0    33.0  
2      AA  N619AA    1141    JFK  MIA     160.0      1089   5.0    42.0  
3      B6  N804JB     725    JFK  BQN     183.0      1576   5.0    44.0  
4      DL  N668DN     461    LGA  ATL     116.0       762   5.0    54.0  
5      UA  N39463    1696    EWR  ORD     150.0       719   5.0    54.0  
6      B6  N516JB     507    EWR  FLL     158.0      1065   5.0    55.0  
7      EV  N829AS    5708    LGA  IAD      53.0       229   5.0    57.0  
8      B6  N593JB      79    JFK  MCO     140.0       944   5.0    57.0  
9      AA  N3ALAA     301    LGA  ORD     138.0       733   5.0    58.0

##Some Tips

* This assignment involves extensive Data frame splitting and aggregation. You should look into the details of the methods groupby, transform, sum, count, mean etc
* Many of the tasks in the assignment can be done either through the Pandas Data Frame or by converting the data frames to Series. Many of the methods in the numpy are applicable to Series only. When stuck, try to explore the type of object (Pandas Data Frame or Numpy Series) you are dealing with.

##Question 1
Let’s explore flights from NYC to Seattle. Use the flights dataset to answer the following questions.

(a) How many flights were there from NYC airports to Seattle in 2013?

In [5]:
dest_seattle = len(flights_df[flights_df['dest'] == 'SEA'])

In [6]:
dest_seattle

3923

Since all the flights originate form NYC, we need to consider only the flights where the destination is Seattle. Here we see that we have performed indexing of only those flights in the NYC dataset whose 'dest' is SEA. Once we do that we calculate the number of rows in the filtered dataset which is 3923

(b) How many airlines fly from NYC to Seattle?

In [7]:
unique_carriers = set(flights_df[flights_df['dest'] == 'SEA']['carrier'])

In [8]:
print unique_carriers, len(unique_carriers)

set(['AA', 'B6', 'AS', 'UA', 'DL']) 5


<b> Ans </b> Here we perform the same filtering as above but also reduce the dataset to only the 'carrier' column. After doing this we pass the dataset through the set datatype in order to obtain unique carrier names. Thus we can see that there are 5 unique carriers from NYC to Seattle are as above. Just for curiosity I took a look at the names of these given acronyms which are American Airlines, Alaskan Airlines, JetBlue, Delta and United Airlines :) 

(c) How many unique air planes fly from NYC to Seattle?

In [12]:
len(set(flights_df[flights_df['dest'] == 'SEA']['flight']))

166

<b> Ans </b> We perform filtering first on the flights dataframe to obtain only those flights that fly from NYC to SEA. After doing that we perform filtering on the 'flight' column that contains flight numbers which are unique for every airplane. This helps us identify the number of unique air planes that fly from NYC to Seattle 

(d) What is the average arrival delay for flights from NC to Seattle?

In [22]:
flights_seattle = flights_df[flights_df['dest'] == 'SEA']
print flights_seattle.head()
which_vals = flights_seattle[np.isnan(flights_seattle['arr_delay'])]
which_vals.head()
which_vals.sum()



     Unnamed: 0  year  month  day  dep_time  dep_delay  arr_time  arr_delay  \
78           79  2013      1    1     724.0       -1.0    1020.0      -10.0   
93           94  2013      1    1     743.0       13.0    1059.0        3.0   
165         166  2013      1    1     857.0        6.0    1157.0      -25.0   
401         402  2013      1    1    1418.0       -1.0    1726.0       -6.0   
404         405  2013      1    1    1421.0       26.0    1735.0       26.0   

    carrier tailnum  flight origin dest  air_time  distance  hour  minute  
78       AS  N594AS      11    EWR  SEA     338.0      2402   7.0    24.0  
93       DL  N3760C     495    JFK  SEA     349.0      2422   7.0    43.0  
165      UA  N45440    1670    EWR  SEA     343.0      2402   8.0    57.0  
401      UA  N37464      16    EWR  SEA     348.0      2402  14.0    18.0  
404      B6  N503JB      83    JFK  SEA     349.0      2422  14.0    21.0  


Unnamed: 0  year  month  day  dep_time  dep_delay  arr_time  \
14685        14686  2013      1   17    1732.0        5.0    2206.0   
15215        15216  2013      1   18    1005.0        5.0    1429.0   
15624        15625  2013      1   18    1750.0       23.0    2208.0   
49138        49139  2013     10   24    1634.0       35.0    2146.0   
49327        49328  2013     10   24    1919.0      -11.0      30.0   
50454        50455  2013     10   25       NaN        NaN       NaN   
56960        56961  2013     11    2     733.0       -2.0    1642.0   
57058        57059  2013     11    2     905.0       -4.0    1420.0   
80210        80211  2013     11   27     854.0        9.0    1409.0   
82069        82070  2013     11   29    1619.0        9.0    2134.0   
90541        90542  2013     12    8       NaN        NaN       NaN   
98771        98772  2013     12   17       NaN        NaN       NaN   
111285      111286  2013     12   31       NaN        NaN       NaN   
112313      112314  2013      2    2     738.0       -2.0       NaN   
113713      113714  2013      2    3       NaN        NaN       NaN   
117978      117979  2013      2    8       NaN        NaN       NaN   
118007      118008  2013      2    8       NaN        NaN       NaN   
118112      118113  2013      2    8       NaN        NaN       NaN   
118241      118242  2013      2    8       NaN        NaN       NaN   
118617      118618  2013      2    9       NaN        NaN       NaN   
118739      118740  2013      2    9       NaN        NaN       NaN   
118780      118781  2013      2    9       NaN        NaN       NaN   
118954      118955  2013      2    9       NaN        NaN       NaN   
141707      141708  2013      3    6       NaN        NaN       NaN   
159363      159364  2013      3   25    2059.0      103.0      49.0   
159432      159433  2013      3   25       NaN        NaN       NaN   
182848      182849  2013      4   19    1639.0       54.0    2125.0   
198110      198111  2013      5    6     726.0        1.0    1022.0   
213073      213074  2013      5   22     724.0       -1.0    1135.0   
215974      215975  2013      5   25     724.0       -1.0    1216.0   
229512      229513  2013      6    8       NaN        NaN       NaN   
243748      243749  2013      6   23       NaN        NaN       NaN   
245705      245706  2013      6   25       NaN        NaN       NaN   
246702      246703  2013      6   26       NaN        NaN       NaN   
258673      258674  2013      7   10     251.0      471.0     804.0   
283833      283834  2013      8    5     839.0       -3.0    1240.0   
287182      287183  2013      8    8    1538.0        9.0    2027.0   
331830      331831  2013      9   25    1552.0       -7.0    2048.0   

        arr_delay carrier tailnum  flight origin dest  air_time  distance  \
14685         NaN      UA  N462UA     738    EWR  SEA       NaN      2402   
15215         NaN      UA  N490UA     499    EWR  SEA       NaN      2402   
15624         NaN      UA  N481UA     738    EWR  SEA       NaN      2402   
49138         NaN      UA  N442UA     375    EWR  SEA       NaN      2402   
49327         NaN      AA  N3BKAA     235    JFK  SEA       NaN      2422   
50454         NaN      DL  N722TW     443    JFK  SEA       NaN      2422   
56960         NaN      DL  N710TW     183    JFK  SEA       NaN      2422   
57058         NaN      UA  N17244    1581    EWR  SEA       NaN      2402   
80210         NaN      UA  N37465    1581    EWR  SEA       NaN      2402   
82069         NaN      UA  N414UA     654    EWR  SEA       NaN      2402   
90541         NaN      AA  N3HTAA     235    JFK  SEA       NaN      2422   
98771         NaN      AA  N3JPAA     235    JFK  SEA       NaN      2422   
111285        NaN      UA     NaN    1151    EWR  SEA       NaN      2402   
112313        NaN      DL  N647DL    1889    JFK  SEA       NaN      2422   
113713        NaN      DL  N647DL    1889    JFK  SEA       NaN      2422   
117978

<b> Ans </b> We first observe that there are 38 values in the flights form NYC to SEA that have arr_delay values as NaN. Therefore inorder to calculate the average arrival delay we can either impute ( discard ) these values. 

(e) What proportion of flights to Seattle come from each NYC airport?

In [ ]:
# Your code here

 -- Write your answer in English here -- 

## Question 2
Flights are often delayed. Consider the following questions exploring delay patterns.

(a) Which date has the largest average departure delay? Which date has the largest average arrival delay?

In [ ]:
# Your code here

 -- Write your answer in English here -- 

(b) What was the worst day to fly out of NYC in 2013 if you dislike delayed flights?


In [ ]:
# Your code here

 -- Write your answer in English here -- 

(c) Are there any seasonal patterns in departure delays for flights from NYC?

In [ ]:
# Your code here

 -- Write your answer in English here -- 

(d) On average, how do departure delays vary over the course of a day?

In [ ]:
# Your code here

 -- Write your answer in English here -- 

## Question 3
    Which flight departing NYC in 2013 flew the fastest?

In [25]:
flights_df['speed'] = flights_df['distance']/(flights_df['air_time']/60)

In [26]:
flights_df.sort(['speed'], ascending = False).head(1)

/anaconda/envs/gl-env/lib/python2.7/site-packages/ipykernel/__main__.py:1: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)
  if __name__ == '__main__':


Unnamed: 0  year  month  day  dep_time  dep_delay  arr_time  \
216447      216448  2013      5   25    1709.0        9.0    1923.0   

        arr_delay carrier tailnum  flight origin dest  air_time  distance  \
216447      -14.0      DL  N666DN    1499    LGA  ATL      65.0       762   

        hour  minute       speed  
216447  17.0     9.0  703.384615

First we created a new column by dividing the distance of the flight by the total air_time ( in hours) and called it speed. After that we sorted the dataframe in descending order and returned the topmost row which we can see is a Delta Airlines flight to Atlanta From LaGuardia Airport in the year 2013 

## Question 4
Which flights (i.e. carrier + flight + dest) happen every day? Where do they fly to?

In [ ]:
# Your code here

 -- Write your answer in English here -- 

## Question 5
Develop one research question you can address using the nycflights2013 dataset. Provide two visualizations to support your exploration of this question. Discuss what you find.


In [ ]:
# Your code here

 -- Enter your discussion here --

## Question 6
What weather conditions are associated with flight delays leaving NYC? Use graphics to explore.

In [ ]:
# Your code here

-- Enter your interpretation here --